lib/fastlane_core/ui/implementations/shell.rb

```diff
@@ -0,0 +1,100 @@
+module FastlaneCore
+  # Shell is the terminal output of things
+  # For documentation for each of the methods open `interface.rb`
+  class Shell < Interface
+    def log
+      return @log if @log
+
+      $stdout.sync = true
+
+      if Helper.is_test?
+        @log ||= Logger.new(nil) # don't show any logs when running tests
+      else
+        @log ||= Logger.new($stdout)
+      end
+
+      @log.formatter = proc do |severity, datetime, progname, msg|
+        string = "#{severity} [#{datetime.strftime('%Y-%m-%d %H:%M:%S.%2N')}]: " if $verbose
+        string = "[#{datetime.strftime('%H:%M:%S')}]: " unless $verbose
+
+        string += "#{msg}\n"
+
+        string
+      end
+
+      @log
+    end
+
+    #####################################################
+    # @!group Messaging: show text to the user
+    #####################################################
+
+    def error(message)
+      log.error(message.red)
+    end
+
+    def important(message)
+      log.warn(message.yellow)
+    end
+
+    def success(message)
+      log.info(message.green)
+    end
+
+    def message(message)
+      log.info(message)
+    end
+
+    def command(message)
+      log.info("$ #{message}".cyan.underline)
+    end
+
+    def command_output(message)
+      actual = (message.split("\r").last || "") # as clearing the line will remove the `>` and the time stamp
+      actual.split("\n").each do |msg|
```

Comment:
Right. That's the sort of thing I am thinking about, though I admit that it is probably premature to worry about in this PR.

In very quick research `s.encode(encoding, universal_newline: true)` might be the way to normalize, but I'm sure others with more Ruby experience than me can say better!